components/brave_extension/extension/brave_extension/background/webDiscoveryProject.ts

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) 2021 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// you can obtain one at http://mozilla.org/MPL/2.0/.
+
+import { App } from '../../../../../vendor/web-discovery-project/build'
+
+declare var window: any
+
+if (App !== undefined) {
+  const APP = new App({
+    version: chrome.runtime.getManifest().version
+  })
+  window.WDP = APP
+
+  const WEB_DISCOVERY_PREF_KEY = 'brave.web_discovery_enabled'
+
+  // Dynamically inject WDP content script so that users with the pref disabled
+  // don't need to pay the loading cost on each page.
+  chrome.tabs.onUpdated.addListener((tabId: number, changeInfo: chrome.tabs.TabChangeInfo, tab: chrome.tabs.Tab) => {
```

Comment:
Load content-script dynamically looks much better, thanks!
But I'm not sure that `chrome.tabs.onUpdated` is the right event.
If the page updates some details (like a title) multiple times can it results in double injection of the script?
Have you looked at `chrome.webNavigation.onCompleted / onDOMContentLoaded`  ? (edited)

Reply:
This needs to check changeInfo and should use that url (if it exists) and not tab.url and definitely not tab.pendingUrl. pendingUrl can still be aborted for various reasons and at that point the old page is still loaded. You must wait for it to be committed.

Reply:
Fixed in https://github.com/brave/brave-core/pull/9381/commits/3e21a792eb403dfd573c7fd05f24f2fb010844e7